examples/miscellaneous/plot_outlier_detection_bench.py

```diff
@@ -3,193 +3,414 @@
 
-2. The ROC curve from :class:`~sklearn.metrics.RocCurveDisplay` is computed
-on the same dataset using the knowledge of the labels.
+1. The ROC curves are computed using knowledge of the ground-truth labels
+and displayed using :class:`~sklearn.metrics.RocCurveDisplay`.
 
+2. The performance is assessed in terms of the ROC-AUC.
 """
 
 # Author: Pharuj Rajborirug <pharuj.ra@kmitl.ac.th>
+#         Arturo Amor <david-arturo.amor-quiroz@inria.fr>
 # License: BSD 3 clause
 
-print(__doc__)
+# %%
+# Dataset preprocessing and model training
+# ========================================
+#
+# Different outlier detection models require different preprocessing. In the
+# presence of categorical variables,
+# :class:`~sklearn.preprocessing.OrdinalEncoder` is often a good strategy for
+# tree-based models such as :class:`~sklearn.ensemble.IsolationForest`, whereas
+# neighbors-based models such as :class:`~sklearn.neighbors.LocalOutlierFactor`
+# would be impacted by the ordering induced by ordinal encoding. To avoid
+# inducing an ordering, on should rather use
+# :class:`~sklearn.preprocessing.OneHotEncoder`.
+#
+# Neighbors-based models may also require scaling of the numerical features (see
+# for instance :ref:`neighbors_scaling`). In the presence of outliers, a good
+# option is to use a :class:`~sklearn.preprocessing.RobustScaler`.
+#
+# The following `fit_predict` function returns the average outlier score of X.
+# In this example we set `n_neighbors` to match the expected number of anomalies
+# `expected_n_anomalies = n_samples * expected_anomaly_frac`. This is a good
+# heuristic as long as the proportion of outliers is not very low, as
+# `n_neighbors` should be at least greater than the number of samples in the
+# less populated cluster (see
+# :ref:`sphx_glr_auto_examples_neighbors_plot_lof_outlier_detection.py`).
+
+from sklearn.neighbors import LocalOutlierFactor
+from sklearn.ensemble import IsolationForest
+from sklearn.preprocessing import (
+    OneHotEncoder,
+    OrdinalEncoder,
+    RobustScaler,
+)
+from sklearn.compose import ColumnTransformer
+from sklearn.pipeline import make_pipeline
+from time import perf_counter
+
+
+def fit_predict(X, model_name, expected_anomaly_frac, categorical_columns=()):
+    tic = perf_counter()
+    if model_name == "LOF":
+        n_samples = X.shape[0]
+        n_neighbors = int(n_samples * expected_anomaly_frac)  # expected_n_anomalies
+        preprocessor = ColumnTransformer(
+            [("categorical", OneHotEncoder(), categorical_columns)],
+            remainder=RobustScaler(),
+        )
+        lof = LocalOutlierFactor(n_neighbors=n_neighbors)
+        model = make_pipeline(preprocessor, lof)
+        model.fit(X)
+        y_pred = model[-1].negative_outlier_factor_
+
+    if model_name == "IForest":
+        ordinal_encoder = OrdinalEncoder(
+            handle_unknown="use_encoded_value", unknown_value=-1
+        )
+        iforest = IsolationForest(random_state=rng)
```

Comment:
Using the global `rng` is not a good practice here. We should pass the `random_state` as an argument.